shell/layouts/default.vue

```diff
@@ -466,29 +358,104 @@ export default {
                     params: {
                       cluster: this.clusterId,
                       group:   groupSlug,
-                    }
+                    },
                   },
-                }
+                },
               ],
               weight: -100,
             });
           } else {
             toAdd.push({
-              name:       `navlink-${ entry.name }`,
-              label:      entry.label,
-              isRoot:     true,
+              name:     `navlink-${ entry.name }`,
+              label:    entry.label,
+              isRoot:   true,
               // This is the item that actually shows up in the nav, since this outer group will be invisible
-              children:   [entry],
-              weight:     -100,
+              children: [entry],
+              weight:   -100,
             });
           }
         }
 
         addObjects(out, toAdd);
+
+        return out;
       }
 
       replaceWith(this.groups, ...sortBy(out, ['weight:desc', 'label']));
-      this.gettingGroups = false;
+
+      return out;
+    },
+  },
+
+  watch: {
+    clusterId(a, b) {
+      if (!isEqual(a, b)) {
+        // Store the last visited route when the cluster changes
+        this.setClusterAsLastRoute();
+      }
+    },
+
+    async currentProduct(a, b) {
+      if (!isEqual(a, b)) {
+        if (a.inStore !== b.inStore || a.inStore !== 'cluster') {
+          const route = {
+            name:   'c-cluster-product',
+            params: {
+              cluster: this.clusterId,
+              product: a.name,
+            },
+          };
+
+          await this.$store.dispatch('prefs/setLastVisited', route);
+        }
+      }
+    },
+
+    $route(a, b) {
+      this.$nextTick(() => this.syncNav());
+    },
+  },
+
+  async created() {
+    await this.$store.dispatch('prefs/setLastVisited', this.$route);
+  },
+
+  mounted() {
+    // Sync the navigation tree on fresh load
+    this.$nextTick(() => this.syncNav());
```

Comment:
If we wanted this to be more declarative, the open/closed group state could also be refactored so that we wouldn't need `syncNav`.